Common/Models/CommandHistory.cs

```diff
@@ -0,0 +1,83 @@
+using System;
+using System.Collections.Generic;
+using Newtonsoft.Json.Linq;
+
+namespace Microsoft.Azure.Devices.Applications.RemoteMonitoring.Common.Models
+{
+    public class CommandHistory
+    {
+        /// <summary>
+        /// Internal serialization deserialization constructor.
+        /// </summary>
+        internal CommandHistory()
+        {
+        }
+
+        /// <summary>
+        /// Creates a new instance of a command history.
+        /// </summary>
+        /// <param name="name">The name of the command.</param>
+        /// <param name="parameters">Dynamic list of parameters issued with the command.</param>
+        public CommandHistory(string name, dynamic parameters = null)
+        {
+            if (string.IsNullOrEmpty(name))
+            {
+                throw new ArgumentNullException("name");
+            }
+
+            Name = name;
+            MessageId = Guid.NewGuid().ToString();
+            CreatedTime = DateTime.UtcNow;
+            SetParameters(parameters);
+        }
+
+        public string Name { get; set; }
+        public string MessageId { get; set; }
+        public DateTime CreatedTime { get; set; }
+        public DateTime UpdatedTime { get; set; }
+        public string Result { get; set; }
+        public string ErrorMessage { get; set; }
+        public dynamic Parameters { get; set; }
+
+        public override string ToString()
+        {
+            return Newtonsoft.Json.JsonConvert.SerializeObject(this);
+        }
+
+        /// <summary>
+        /// Adds the entire set of parameters to the command history.
+        /// </summary>
+        /// <param name="parameters">The parameters to set on the history. Must be either a JArray, JObject or a <see cref="Dictionary"/> instance.</param>
+        public void SetParameters(dynamic parameters)
+        {
+            if (parameters == null)
+            {
+                return;
+            }
+
+            if (parameters.GetType() == typeof(Dictionary<string, object>))
+            {
+                var newParam = new JObject();
+                foreach (var parameter in ((Dictionary<string, object>) parameters))
+                {
+                    newParam.Add(parameter.Key, new JValue(parameter.Value));
+                }
+                Parameters = newParam;
+            }
+        }
```

Comment:
Well that goes against the first check that makes sure that if the parameter is null it should simply return. I'm not sure what the usage intent of this is yet so gonna leave it as is. If the intent is that Set can be used to clear out the parameters then I agree. Otherwise this method will throw if parameters passed is null. Will need more context before I make that call, so keeping it as is for now.